pkg/archive/zip.go

```diff
@@ -68,40 +70,156 @@ func (e ZipEntry) WriteTo(zw *zip.Writer) error {
 	return nil
 }
 
-// StabilizeZip strips volatile metadata and rewrites the provided archive in a standard form.
-func StabilizeZip(zr *zip.Reader, zw *zip.Writer) error {
-	defer zw.Close()
-	var ents []ZipEntry
-	for _, f := range zr.File {
-		r, err := f.Open()
+// MutableZipFile wraps zip.File to allow in-place modification of the original.
+type MutableZipFile struct {
+	zip.FileHeader
+	File       *zip.File
+	mutContent []byte
+}
+
+func (mf *MutableZipFile) Open() (io.Reader, error) {
+	if mf.mutContent != nil {
+		return bytes.NewReader(mf.mutContent), nil
+	}
+	return mf.File.Open()
+}
+
+func (mf *MutableZipFile) SetContent(content []byte) {
+	mf.mutContent = content
+}
+
+// MutableZipReader wraps zip.Reader to allow in-place modification of the original.
+type MutableZipReader struct {
+	*zip.Reader
+	File []*MutableZipFile
+}
+
+func NewMutableReader(zr *zip.Reader) MutableZipReader {
+	mr := MutableZipReader{Reader: zr}
+	for _, zf := range zr.File {
+		mr.File = append(mr.File, &MutableZipFile{File: zf, FileHeader: zf.FileHeader})
+	}
```

Comment:
Does the MutableReader need to keep the underlying reader if it already has an in-memory copy of all the files?

Reply:
it's a bit tricky. While it doesn't need to, there are a number of methods that the original exports that we get to inherit by retaining the pointer. and given the name "MutableZipReader", departing from the actual supported API of the reader would be suboptimal and so would be better served by changing the name. MutableZipFile re-implements the original API in spite of not holding on to the original pointer.

all things considered, I don't think it's terribly valuable removing it but it's not strictly required by the stabilizers.